app/controllers/DataUploadController.scala

```diff
@@ -77,37 +81,45 @@ class DataUploadController @Inject()(sessionService: SessionService,
-    implicit request =>
+    implicit request: Request[JsValue] =>
+      Logger.warn("receiving request on V1 processCsvFileDataFromFrontend")
       val startTime =  System.currentTimeMillis()
       request.body.validate[UpscanCsvFileData].fold(
         valid = res => {
           val schemeInfo: SchemeInfo = res.schemeInfo
           Logger.debug("SCHEME TYPE: " + schemeInfo.schemeType)
           deliverFileProcessingMetrics(startTime)
-          Future.sequence(process(res.callbackData, empRef)(hc, schemeInfo, request)).flatMap { result =>
-            val totalRowCount = result.foldLeft(0) ((accum,inputTuple) => accum +inputTuple._2)
-            sessionService.storeCallbackData(res.callbackData.head, totalRowCount).map {
-              case callback: Option[UpscanCallback] if callback.isDefined =>
-                val numberOfSlices = result.map(_._1).sum
-                Ok(numberOfSlices.toString)
-              case _ => Logger.error(s"csv storeCallbackData failed with Exception , timestamp: ${System.currentTimeMillis()}.")
-                throw ERSFileProcessingException("csv callback data storage in sessioncache failed ", "Exception storing csv callback data")
-            } recover {
-              case e:Throwable => Logger.error(s"csv storeCallbackData failed with Exception ${e.getMessage}, timestamp: ${System.currentTimeMillis()}.")
-                throw e
-            }
-          } recover {
-            case e: ERSFileProcessingException => {
-              Logger.error(e.message)
+
+          val processedFiles: List[Future[Either[Throwable, CsvFileContents]]] = processCsvService.processFiles(res, streamFile)
+
+          val extractedSchemeData: Seq[Future[Either[Throwable, (Int, Int)]]] = processedFiles.map{ list =>
+            list.flatMap(processCsvService.extractSchemeData(res.schemeInfo, empRef, _))}
+
+          Future.sequence(extractedSchemeData).flatMap{ oneFileResults => oneFileResults.find(_.isLeft) match {
+            case Some(Left(throwable: ERSFileProcessingException)) =>
+              Logger.error(s"[DataUploadController][processCsvFileDataFromFrontend] ERS file processing exception: ${throwable.message}")
               deliverFileProcessingMetrics(startTime)
-              Accepted(e.message)
-            }
-            case er => {
-              Logger.error(er.getMessage, er)
+              Future(Accepted(throwable.message))
+            case Some(Left(throwable)) =>
+              Logger.error(s"[DataUploadController][processCsvFileDataFromFrontend] Unknown exception: ${throwable.getMessage}, full exception: $throwable")
               deliverFileProcessingMetrics(startTime)
-              InternalServerError
-            }
-          }
+              Future(InternalServerError)
+            case None =>
+              val result: Seq[(Int, Int)] = oneFileResults.map(_.right.get)
+              val totalRowCount = result.foldLeft(0) ((accum,inputTuple) => accum + inputTuple._2)
```

Comment:
Following on from our conversation... if we know that this Seq[(Int, Int)] is only going to have a length of one (since we're already dealing with a single file), should we not just take noOfRows directly from our sequence entry? 

Same goes for the sum of slices further down.